control-plane/controllers/configentries/terminatinggateway_controller.go

```diff
@@ -41,6 +116,221 @@ func (r *TerminatingGatewayController) UpdateStatus(ctx context.Context, obj cli
+		if strings.Contains(role.Name, termGW.Name) {
+			terminatingGatewayRoleID = role.ID
+			break
+		}
+	}
+
+	if terminatingGatewayRoleID == "" {
+		return errors.New("terminating gateway role not found")
+	}
+
+	termGwRole, _, err := client.ACL().RoleRead(terminatingGatewayRoleID, nil)
+	if err != nil {
+		return err
+	}
+
+	var termGWPolicy *capi.ACLRolePolicyLink
+
+	for _, policy := range termGwRole.Policies {
+		if strings.Contains(policy.Name, termGW.Name) {
+			termGWPolicy = policy
+			break
+		}
+	}
+
+	var termGWPoliciesToKeep []*capi.ACLRolePolicyLink
+	var termGWPoliciesToRemove []*capi.ACLRolePolicyLink
+
+	existingTermGWPolicies := mapset.NewSet[string]()
+
+	for _, policy := range termGwRole.Policies {
+		existingTermGWPolicies.Add(policy.Name)
+	}
+
+	if termGW.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("terminating gateway is being modified")
+		termGWPoliciesToKeep, termGWPoliciesToRemove, err = r.handleModificationForPolicies(log, client, existingTermGWPolicies, termGW.Spec.Services)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Info("terminating gateway is deleted")
+		termGWPoliciesToKeep, termGWPoliciesToRemove = handleDeletionForPolicies(termGW.Spec.Services)
+	}
+
+	termGWPoliciesToKeep = append(termGWPoliciesToKeep, termGWPolicy)
+	termGwRole.Policies = termGWPoliciesToKeep
+
+	log.Info("updating acl role")
+
+	_, _, err = client.ACL().RoleUpdate(termGwRole, nil)
+	if err != nil {
+		return err
+	}
+
+	err = r.conditionallyDeletePolicies(log, client, termGWPoliciesToRemove, termGW.Name)
+	if err != nil {
+		return err
+	}
+
+	log.Info("finished updating acl roles")
+	return nil
+}
+
+func handleDeletionForPolicies(services []v1alpha1.LinkedService) ([]*capi.ACLRolePolicyLink, []*capi.ACLRolePolicyLink) {
+	var termGWPoliciesToRemove []*capi.ACLRolePolicyLink
+	for _, service := range services {
+		termGWPoliciesToRemove = append(termGWPoliciesToRemove, &capi.ACLRolePolicyLink{Name: servicePolicyName(service.Name)})
+	}
+	return nil, termGWPoliciesToRemove
+}
+
+func (r *TerminatingGatewayController) handleModificationForPolicies(log logr.Logger, client *capi.Client, existingTermGWPolicies mapset.Set[string], services []v1alpha1.LinkedService) ([]*capi.ACLRolePolicyLink, []*capi.ACLRolePolicyLink, error) {
+	// add one to length to include the terminating-gateway policy for itself
+	termGWPoliciesToKeep := make([]*capi.ACLRolePolicyLink, 0, len(services)+1)
+	termGWPoliciesToRemove := make([]*capi.ACLRolePolicyLink, 0, len(services))
+
+	termGWPoliciesToKeepNames := mapset.NewSet[string]()
+	for _, service := range services {
+		var data bytes.Buffer
+		if err := servicePolicyTpl.Execute(&data, templateArgs{
+			EnableNamespaces: r.NamespacesEnabled,
+			Namespace:        defaultIfEmpty(service.Namespace),
+			ServiceName:      service.Name,
+		}); err != nil {
+			// just panic if we can't compile the simple template
+			// as it means something else is going severly wrong.
+			panic(err)
+		}
```

Comment:
#nit I _think_ this could be moved down into the `if existingPolicy == nil` block